src/nupic/types/Serializable.hpp

```diff
@@ -61,22 +119,145 @@ class Serializable {
   virtual void save(std::ostream &stream) const = 0;
   virtual void load(std::istream &stream) = 0;
 
+
+  virtual inline void saveToFile_ar(std::string filePath, SerializableFormat fmt=SerializableFormat::BINARY) const {
+      std::string dirPath = Path::getParent(filePath);
+	  Directory::create(dirPath, true, true);
+	  std::ofstream out(filePath, std::ios_base::out | std::ios_base::binary);
+	  out.exceptions(std::ofstream::failbit | std::ofstream::badbit);
+	  out.precision(std::numeric_limits<double>::digits10 + 1);
+	  out.precision(std::numeric_limits<float>::digits10 + 1);
+		saveToStream_ar(out, fmt);
+		out.close();
+	}
+  virtual inline void saveToStream_ar(std::ostream &out, SerializableFormat fmt=SerializableFormat::BINARY) const {
+    arWrapper arw;
+    arw.fmt = fmt;
+		switch(fmt) {
+		case SerializableFormat::BINARY:  { 
+      cereal::BinaryOutputArchive ar(out);
+      arw.binary_out = &ar;
+      cereal_adapter_save(arw); 
+    } break;
+		case SerializableFormat::PORTABLE: {
+      cereal::PortableBinaryOutputArchive ar( out, cereal::PortableBinaryOutputArchive::Options::Default() ); 
+      arw.portable_out = &ar;
+      cereal_adapter_save(arw); 
+    } break;
+		case SerializableFormat::JSON: {
+      cereal::JSONOutputArchive ar(out);
+      arw.json_out = &ar;
+      cereal_adapter_save(arw);
+    } break;
+		case SerializableFormat::XML: {
+      cereal::XMLOutputArchive ar(out);
+      arw.xml_out = &ar;
+      cereal_adapter_save(arw);
+    } break;
+		default: NTA_THROW << "unknown serialization format.";
+		}
+  }
+  
+  virtual inline void loadFromFile_ar(std::string filePath, SerializableFormat fmt=SerializableFormat::BINARY) {
+    std::ifstream in(filePath, std::ios_base::in | std::ios_base::binary);
+    in.exceptions(std::ifstream::failbit | std::ifstream::badbit);
+		loadFromStream_ar(in, fmt);
+		in.close();
+	}
+	virtual inline void loadFromStream_ar(std::istream &in,  SerializableFormat fmt=SerializableFormat::BINARY) {
+    arWrapper arw;
+    arw.fmt = fmt;
+		switch(fmt) {
+		case SerializableFormat::BINARY: {
+      cereal::BinaryInputArchive ar(in);
+      arw.binary_in = &ar;
+      cereal_adapter_load(arw);
+    } break;
+		case SerializableFormat::PORTABLE: {
+      cereal::PortableBinaryInputArchive ar(in, cereal::PortableBinaryInputArchive::Options::Default());
+      arw.portable_in = &ar;
+      cereal_adapter_load(arw);
+    } break;
+		case SerializableFormat::JSON: {
+      cereal::JSONInputArchive ar(in);
+      arw.json_in = &ar;
+      cereal_adapter_load(arw);
+    } break;
+		case SerializableFormat::XML: {
+      cereal::XMLInputArchive ar(in);
+      arw.xml_in = &ar;
+      cereal_adapter_load(arw);
+    } break;
+		default: NTA_THROW << "unknown serialization format.";
+		}
+  }
+
+  // Note: if you get a compile error saying this is not defined,
+  //       or that "cannot instantiate abstract class"
+  //       add the macro 'CerealAdapter' in the derived class.
```

Comment:
as my note above, 
if we add the macro here, would it automatically be in all the classes that include Serializable? 
If not, add a comment to Dimensions at the line with CerealAdapter to look here in Serializable for explanation.

Reply:
> If not, add a comment to Dimensions at the line with CerealAdapter to look here in Serializable for explanation.

What it does is add two methods to every derived class that uses Serializable as its parent so that inheritance can be used by the parent class.  I will add some explanation to Serializable.